### ISSUE TITLE
global default options and handlers

### ISSUE THREAD

**Issue by farskid** [developit/unfetch#22]
Since fetch apis are used a lot in a typical web application, it would be nice to have a `unfetch.defaults` object that sets default parameters of all requests.
A use case would be to set base URL of all requests.
```
unfetch.defaults.baseURL = 'my-api-endpoint'
```

**Comment by tunnckoCore** [comment]
Sounds good, but don't believe we can get 500b, haha

**Comment by farskid** [comment]
Well, having a useful feature to cost some bytes wouldn't harm I guess.
Let's watch the package size to see if there is a chance for adding this feature.

**Comment by tunnckoCore** [comment]
I'm agree with you and was thinking the same thing when `mitt` came  - in its case we can get few more features if we allow 20-30bytes more, but @developit is holding strongly that it should be 200b. I'm agree with him too and there was a discussion like if we allow such things we'll get to point such as _"let's add more more 1-2kb"_ haha and was mentioned `momentjs` that it started like micro lib.

So yea, it is good thing to try to be strict.

**Comment by farskid** [comment]
@tunnckoCore HAHA did not know that `momentjs` was a micro library at the beginning. Looking forward to hear about the decision from @developit . This feature whether added or not, both seems logical to me at the moment. :)

**Comment by tunnckoCore** [comment]
Yea, that's the exact comment https://github.com/developit/mitt/pull/11#issuecomment-273132616, and i'm totally agree with him.

**Comment by developit** [comment]
This discussion makes me happy :)
Let's keep this on a list of nice-to-haves that we can look to if we save bytes. 

Btw - for defaults, we should make a wrapper library, that way they would work for native fetch too 😀

**Comment by farskid** [comment]
@developit You mean like an IIFE that wraps the native or unfetch altogether?

**Comment by developit** [comment]
Yup, a way to inject defaults by proxying window.fetch:

```js
let real = window.fetch;
window.fetch = (url, options) => {
  // Apply URL and options transforms here
  return old(url, ootions)
    .then(options.after || Object);
               // ^ middleware!!
};
```

**Comment by farskid** [comment]
Yeah, good pattern, agreed.

**Comment by tunnckoCore** [comment]
Absolutely agree to be separate kinda nanolib. Hm, let's coin `nano frameworks`? haha

**Comment by developit** [comment]
Unlibs

**Comment by developit** [comment]
@tunnckoCore @farskid - what do we want to do with these types of issues/PRs? We can't merge them because of the nature of this library (it being a polyfill and thus bound by the spec it emulates), but I hate closing them when there's good discussion and tasks to be taken away.  Maybe we need a repo/place where these ideas/implementations go so that we can build up a set of feature requests for wrapper libraries?

**Comment by farskid** [comment]
@developit @tunnckoCore I guess a wrapper for addon features is a nice idea. Devs can either use the super light version or pay the price of not so much light version for the extra features such as global configurations, extra handlers, ...

**Comment by developit** [comment]
Exactly! And the wrapper will work regardless of whether unfetch or native fetch gets used.

**Comment by farskid** [comment]
Yup so much like a higher order API. Would love to contribute to the wrapper.

**Comment by tunnckoCore** [comment]
Exactly, that's the cool of micro things. : )

**Comment by developit** [comment]
Prior art:
https://github.com/RickWong/fetch-plus

**Comment by iamale** [comment]
I've written [this](https://gist.github.com/iamale/15996ddd1fc3cc15ca094ae2e53a01c9) in one of my projects sometime ago. `fetch-plus` seems a lot nicer though.

**Comment by developit** [comment]
@iamale totally stealing that

**Comment by MobiusHorizons** [comment]
Learned a lot just reading this thread!! There are a lot of good ideas happening here. Is there a way it can be preserved? If it were closed, it could still be linked to from a wiki or something.

**Comment by developit** [comment]
Heh - I think that's why we haven't closed it. Need to find a good home for these things, maybe in the wiki.

**Comment by iamale** [comment]
Another thing that I had in mind since I wrote [arequests] is [this](https://gist.github.com/iamale/09a2a0921db416a8ffb92c56cfd26717):

```js
const api = Afetch({url: "https://api.github.com"});
let mdo = await api.users["mdo"].get();
let json = await mdo.json();  // or whatever you'd do with a regular Fetch object
```

[arequests]: https://github.com/iamale/arequests
